Miscellaneous extensions of notations (including granting BZ5585)

herbelin opened the pull request:
Hi,

This PR is about miscellaneous extensions of the notation mechanism. It includes various fixes, features, mini-documentation and restructuration.

The main extensions are 
- Recursive notations now supports patterns with several occurrences of the recursive term or binder, possibly mixing terms and binders, possibly in reverse left-to-right order, with binders accepting or-patterns. Here are examples of the new possibilities:

  - Mixing several occurrences in terms and binders
    ```coq
    Notation "'exists_non_null' x .. y  , P" :=
      (ex (fun x => x <> 0 /\ .. (ex (fun y => y <> 0 /\ P)) ..))
      (at level 200, x binder).
    ```
  - right-to-left order
    ```coq
    Notation "[ a , .. , b |- A ]" := (cons b .. (cons a nil) .., A).
    ```
  - or-pattern in 'pat
    ```coq
    Check fun '(((x,y),true)|((x,y),false)) => x+y.
    ```

- a new entry `id is pattern` can be used to parse subterms of a notation as patterns, thus granting the wish of @JasonGross and @tchajed of notations with [BZ#5585](https://coq.inria.fr/bugs/show_bug.cgi?id=5585) ("allow binder syntax in non-recursive binder notation").

  - Extension of the syntax for subset types:
    ```coq
    Check {(x,y)|x+y>0}.
    ```
  - A BZ#5585-like example (note that a quote `'` is not needed, but that we may would like to require one in practice for consistency of style).
    ```coq
    Parameter myex : forall {A}, (A -> Prop) -> Prop.
    Notation "'myexists' x , p" := (myex (fun x => p))
      (at level 200, x pattern, p at level 200, right associativity).
    Check myexists (x,y), x>y.
    ```
  - A user-side form of "if t is pat then u else v"; but printing still to do (this is related to discussion at #978)
    ```coq
    Notation "'if' c 'is' p 'then' u 'else' v" :=
      (match c with p => u | _ => v end)
      (at level 200, p pattern).
    ```

See commits for more details.

This is still work in progress, with still further improvements in sight but I have no visibility on when I'll be able to continue working on it next, so I'm publishing it now for receiving a first feedback.

Note that it builds on top of #965 (ocamldebug fix), #978 (improving printing of `match` clauses), #980 (more functions in List), #981 (misc. fixes of notations). [Edit: as well as of #916 (fix for bug BZ#5608)]

Side comment about the API for notations: I feel that it exposes too much of the details of the infrastructure for notation (including things liable to change easily). One would need to think about it.

With respect to plugins/ssr, I made what seems to me the minimalistic changes, hoping I did not introduce errors. In the process, I realize that `'pat` is not well supported in forward tactics. Hints on the intents of Ssreflect developers to improve on this welcome.

JasonGross commented:
> note that a quote `'` is not needed, but that we may would like to require one in practice for consistency of style

Hmm, I'm not sure what to think about this one.  In some sense, Coq is already inconsistent: `let (a, b) := v` means something different from `match v with (a, b) => ...` (which itself means `let '(a, b) := v`)....  So while there's a clear meaning for having `'`, it's not required everywhere (namely, in `match`, where it is in fact forbidden).  In some sense, it's a shame that the meaning of `'` isn't reversed, because that would grant consistency between `let` and `match`, and then you could even generalize `match` to permit `match v with '(a, b) => k end` which would be equivalent to what is now written as `let (a, b) := v in k`.  I think this would be my preferred syntax, possibly hidden behind some sort of compatibility frob for a couple of versions.  (This would require eventual changes to developments making intentional use of `let '` for constructors other than `pair`, and hopefully there can be a tool that makes the change for us.)  I think this is beyond the scope of this PR, though.

herbelin commented:
@JasonGross: I agree with your analysis. I'm not a big fan of `let (x,...,y) := ...` (even though it is probably me who implemented it). I prefer `let `(x,...,y) := ...` which supports the whole language of patterns and is more consistent with `match`.

One problem I see for dropping `'` is in `fun` and `forall`. Since we can write `fun (x y:nat) => ...` or `fun x y => ...`, how would we distinguish it from `fun (S n:nat) =>` or `fun S n => ...`?

JasonGross commented:
Oh, yeah, that's an issue.  Perhaps let the notation authors decide whether to allow, require, or forbid `'`, with some sort of syntax?  (E.g., pattern, pattern with quote, pattern without quote?)

JasonGross commented:
Hmm, so, one way to distinguish it is that the following seems to be one way of consistently disambiguating:
- `fun S n =>` two arguments
- `fun (S n) =>` pattern
- `fun (S n : nat) =>` two arguments
- `fun ((S n) : nat) =>` pattern

herbelin commented:
@JasonGross: Technically, this would be a bit tricky to support, since patterns can come equipped with any arbitrary notation, but accepting to renounce to pattern notations starting with `{` or `` ` ``, and with a special treatment for `(t,...,u)` so that it factorizes with `(x y : nat)` (i.e. some special treatment to factorize `"("; t = constr; ","; LIST1 constr; ...` and `"("; x = ident; LIST1 ident; ...`, using camlp4/p5 syntax), that could work.

herbelin commented:
Hi, while being under way, I started working at the ability to add new "subentries" for terms, allowing  things like:
```coq
Notation "[ x ]" := x (x custom myconstr at level 6).
Notation "x + y" := (add x y) (in custom myconstr, at level 5).
Notation "x * y" := (mul x y) (in custom myconstr, at level 4).
Notation "< x >" := x (in custom myconstr, at level 3, x constr).
```
This is a long-standing request, @ggonthier would have liked to have it for instance, but also @beta-ziliani. This was re-discussed recently at POPL with @ejgallego, @beta-ziliani and others. Maybe someone has actually already started something about that?

I'm otherwise wondering about the appropriate user interface for it.
- For instance, do we need levels, or being able to have `myconstr4`, `myconstr5` and managing levels by hands would be enough? If I'm asking, it is I feel that levels would be better but this has some implementation cost due to camlp5 not supporting backtracking for levels, and the backtracking having to be managed coq side. At the same time, managing backtracking of levels is already done for constr, so it would just be needed to generalize the mechanism a little bit.
- When or where do we need subentries? The typical examples I have in mind are for describing grammar of subexpressions, such as, say a language `e  ::=  (e)  ||  e|e  ||  e*  ||  e+  || l`, etc., where scopes are not sufficient, because the symbols have different priority or arity, or interfere with other symbols, etc. Are there other typical examples?
- I'm wondering about how to implement printing for such subentries? My current view is that we should give a special status to rules coercing an entry to another (such as `Notation "[ x ]" := x (x custom myconstr).` which embeds `myconstr` in `constr`. Then we shall look for patterns up to coercions, i.e. the map from `constr` to `notation` (`uninterp_notation`) would be extended into a map from `term * grammar_entry` to `notation * grammar_coercion list` or something like that. Not very complicated but not trivial either, so I'm wondering if we could find a simpler approach or if there is an inherent complexity in these coercions and that we cannot escape giving them a special status.

beta-ziliani commented:
@herbelin this is great! Is there a way to specify a different entry for a subterm? Like `x` in this example:

    Notation "< x >" := x (in custom myconstr, at level 3, x in custom anotherconstr at level 10).

herbelin commented:
@beta-ziliani: yes. Here is an example with three entries:
```coq
Notation "[ x ]" := x (x custom myconstr at level 6).
Notation "< x >" := x (in custom myconstr, at level 3, x custom anotherconstr at level 10).
Notation "# x" := (Some x) (in custom anotherconstr, at level 8, x constr at level 9).
Check [ < # 0 > ].
```

herbelin commented:
One of the reasons for incompatibility in that branch is that `{ x | P }` now extends to `{ pat | P }` and there are a couple of tested scripts which do things like `{ I | P }` or `{ pair | P }` where `I` and `pair` get consequently interpreted as constructors while they were expected to be variables.

So, this raises the question: shall we prefer to extend the syntax of `{ x | P }` into `{ pat | P }` or shall we prefer the compatibility?

An intermediate solution would be to accept only `{ 'pat | P }`... and this is not unrelated to the discussion [above](https://github.com/coq/coq/pull/982#issuecomment-323417432) and [there](https://github.com/coq/coq/pull/1003#issuecomment-326114564).

Any opinions?

In any cases, to let the second alternative possible, I propose to implement one or both of these new entry modifiers:
- `id is constr as ident` meaning that `id` should be parsed as a `constr` (for factorization), but only accepted if an `ident`;
- `id is constr as pattern` meaning that `id` should be parsed as a `constr`, but only accepted if a `pattern`.

The former was the semantics by default as soon as `id` was bound in the right hand side, but with the arrival of binding patterns, it is the latter which is the semantics by default in this PR.

So, second choice to answer: do we preserve compatibility by keeping the first semantics by default, or do we prefer what maybe is more natural which is to have the second semantics by default? (Again, this is not unrelated to the question of `pat` vs `'pat`.)

RalfJung commented:
Feature-wise, I think this is great and should satisfy our needs for controlling the grammar used for parsing. Thanks a lot @herbelin! :)

One thing I am wondering is which limitations there are for the "grammar-changing" notation. In your examples, they always have a pair of patching parenthesis around them; is that necessary? Or could be have something like
```
Notation "x '|-' y" := (uPred_entails x y) (x, y custom myconstr at level 6).
```

How do these custom grammars interact with scopes? Are they entirely orthogonal, or do scopes only apply to the main grammar, or is there a shared scope stack, or something else?

Also, I think the syntax `in custom` is not very self-explaining. Custom what? Ideally, `in` should be followed by a noun that's somehow describing this feature (and hence will also appear in the docs and be googleable). Maybe something like `in grammar` or `in table`.

Is it possible to define an Ltac Notation that has one argument parsed in a different table? (We have wanted that for scopes frequently, also see <https://coq.inria.fr/bugs/show_bug.cgi?id=4381>.)

@robbertkrebbers IIRC we wanted to use this to represent proof mode intro patterns as a datatype rather than a string. However, I expect actually experimenting with this is also blocked on finding a way to use Coq's binders rather than strings at the "leaves" of the pattern?

robbertkrebbers commented:
> @robbertkrebbers IIRC we wanted to use this to represent proof mode intro patterns as a datatype rather than a string. However, I expect actually experimenting with this is also blocked on finding a way to use Coq's binders rather than strings at the "leaves" of the pattern?

Yes, indeed, we want this feature to represent introduction patterns. Not just in Iris proof mode, but also in MetaCoq. The remaining issue is indeed how to deal with identifiers at the leafs, we after all still do not want to write `["x" "y"|"x"]` in our introduction patterns.

herbelin commented:
@RalfJung, @robbertkrebbers:
> One thing I am wondering is which limitations there are for the "grammar-changing" notation. In your examples, they always have a pair of patching parenthesis around them; is that necessary? Or could be have something like
> ```
>  Notation "x '|-' y" := (uPred_entails x y) (x, y custom myconstr at level 6).
> ```
There is no requirement about having parentheses, and, yes, you could do something as you describe. Note however that the above would be fine for the printer but not for the parser. By default the parser goes from left to right, so it cannot know in advance that it has to parse a `myconstr` until it eventually finds a `|-`. So, for the parser, a mark has to exist on the left to indicate when a custom entry has to be parsed. For instance, the following would be ok:
```
Notation "[ x '|-' y ]" := (uPred_entails x y) (x, y custom myconstr at level 6).
Notation "# x '|-' y" := (uPred_entails x y) (x, y custom myconstr at level 6).
```

> How do these custom grammars interact with scopes? Are they entirely orthogonal, or do scopes only apply to the main grammar, or is there a shared scope stack, or something else?

This is a good question and I still don't know well what should be the good interaction. We need some experimentation here. Currently, there is a shared scope stack, if I remember correctly what I did.

> Also, I think the syntax `in custom` is not very self-explaining. Custom what? Ideally, `in` should be followed by a noun that's somehow describing this feature (and hence will also appear in the docs and be googleable). Maybe something like `in grammar` or `in table`.

There are two kinds of information to give. What are the entry and level of the subentries and what is the entry and level of the main entry. For subentries, the usual syntax is `id at level N` to give a level, `id ident` to give a subentry name, and, since this PR, there is also `id pattern at level N` to give both at the same time.

I wanted to mimic this for the main entry. For the main entry, we already have `at level N`, so it could be just generalized into `myconstr at level N` (similarly to `id pattern at level N` but without the `id`). But this would mean a syntax starting with an arbitrary entry name while we already have the `id ident` syntax accepting an arbitrary leading name. This would be a source of confusion, so, a discriminating keyword was needed. It could be only `custom` (or whatever similar) but I did not find it explicit enough so I used `in custom`. Any other proposal can be considered, but I would find a `in table` or `in grammar` a bit too much different from the current `at level N` which implicitly means `in constr at level N`). Maybe instead `is myconstr at level N`, or `named myconstr at level N`...

> Is it possible to define an Ltac Notation that has one argument parsed in a different table? (We have wanted that for scopes frequently, also see https://coq.inria.fr/bugs/show_bug.cgi?id=4381.)

In theory, yes, but I just checked and it does not work. This is a problem of syntax about which maybe @ppedrot would be able to help: `Tactic Notation` currently accepts levels only for `tactic`, and this is treated in an ad hoc way. So, `Tactic Notation` does not accept to write neither `Tactic Notation "mytac" myconstr2(x) := ...` (because `myconstr2` is recognized as an entry named `myconstr2` and not as `myconstr` a level 2), nor `Tactic Notation "mytac" myconstr(x) := ...` (because `myconstr` expects a level). One would typically need a more robust syntax such as `Tactic Notation "mytac" myconstr(x,at level 2) := ...` or similar. Or to be able to declare a default level for `myconstr`. With also the question of what @ppedrot wants to do in the context of Ltac2, etc.

> Yes, indeed, we want this feature to represent introduction patterns. Not just in Iris proof mode, but also in MetaCoq. The remaining issue is indeed how to deal with identifiers at the leafs, we after all still do not want to write ["x" "y"|"x"] in our introduction patterns.

Indeed, parsing strings as identifiers is not provided by this PR. @roglo had actually the same need (in order to represent named binders in first-order logic), so maybe something can be done with him. A typical user-level API could be a modifier of the form `x ident as string` to tell that the subentry has to be parsed as an ident but then translated to a string. This is something that the `string_syntax_plugin` plugin should provide. This would be a bit of work but if we know what API we exactly want, this might not be too difficult.

ppedrot commented:
> With also the question of what @ppedrot wants to do in the context of Ltac2, etc.

Not sure if relevant, but in Ltac2 scopes are described by a kind of S-expressions, and some scopes like `tactic` accept an integer argument to describe the level. You do have to implement them ML-side, but scopes are easily composable on Coq-side, see e.g. the master-level trickery [used to implement Ltac1 notations](https://github.com/ppedrot/ltac2/blob/master/theories/Notations.v).

RalfJung commented:
> It could be only custom (or whatever similar) but I did not find it explicit enough so I used in custom. Any other proposal can be considered, but I would find a in table or in grammar a bit too much different from the current at level N which implicitly means in constr at level N). 

How is `in custom <name> at level <N>` more similar to `at level <N>` than `in grammar <name> at level <N>`?

> A typical user-level API could be a modifier of the form x ident as string to tell that the subentry has to be parsed as an ident but then translated to a string. 

That would be a great start!
However, that on its own would not be sufficient as Ltac still lacks a way to convert strings to idents. These intro patterns can refer to both Coq-level and Iris-level hypothesis, so if all the idents are turned into strings before being passed to the ltac code, then that code would need a way to convert those strings to idents in case this name goes into the Coq context. (Also see <https://github.com/coq/coq/pull/285>.)

ppedrot commented:
@RalfJung Sorry to hijack this thread, but you may be interested into having a look at the current Ltac2 implementation (ppedrot/ltac2@4db5aab79ccb8007a5186bdfaefa3ec7421b04e4). There is no official release yet, but you can already mix Ltac1 and Ltac2 code in a typical proof, with a few limitations, see the doc. It would probably help you to work around the string-to-ident issue in a perennial way.

herbelin commented:
> How is in custom <name> at level <N> more similar to at level <N> than in grammar <name> at level <N>?

Well, ok. What about `in entry <name>` and `in entry <name> at level <N>` then? I found "grammar" a bit too general term and the reference manual is already using "entry".

> These intro patterns can refer to both Coq-level and Iris-level hypothesis, so if all the idents are turned into strings before being passed to the ltac code, then that code would need a way to convert those strings to idents in case this name goes into the Coq context. (Also see #285.)

So, if I understand correctly, you don't know at parsing time whether it'll be a Coq ident or an Iris ident. Then, I can suspect that conversion in both ways is indeed necessary.

I forgot about #285, thanks for reminding it. Aren't the ideas in @ppedrot's [string-to-ident](https://github.com/ppedrot/coq-string-ident) plugin, or variations about them, enough for your needs then? (I guess that @ppedrot is talking about parsing all of Coq and Iris idents as "first-order" tactic language objects, then turn afterwards those recognized as Iris identifiers to string using the plugin??)

RalfJung commented:
> What about in entry <name> and in entry <name> at level <N> then? I found "grammar" a bit too general term and the reference manual is already using "entry".

Indeed it does. In chapter one, it says

> The notation “entry sep … sep entry” stands for a non empty sequence of expressions parsed by entry and separated by the literal “sep”1.

However, the term "entry" is not explained. I have to say that despite using Coq for years, I had no idea that Coq uses "entry" as the term for a parsing table. When reading `in entry <name>`, I would still not know which kind of a thing `<name>` is an entry in. `entry` is not very descriptive, but well, at least it is consistent.

> So, if I understand correctly, you don't know at parsing time whether it'll be a Coq ident or an Iris ident. 

Correct.

> Aren't the ideas in @ppedrot's string-to-ident plugin, or variations about them, enough for your needs then?

They are, and indeed that plugin would be extremely useful independently of the notation extensions discussed here. Namely, we could just keep using strings for intro patterns and then convert them to Coq idents as needed. We currently don't use that plugin because we have not yet figured out a way for users of Iris on Windows to get it. (From what I hear, using opam is not an option there.)

With the extensions of this PR and `ident as string` and the plugin, we could furthermore get rid of strings in the tactic language (and of our own parser written in Ltac). And finally, if there was a way to get the inverse of `ident as string` for printing notatuions (i.e., print a string without quotes), we could even get rid of strings in the proof state.

herbelin commented:
> However, the term "entry" is not explained. I have to say that despite using Coq for years, I had no idea that Coq uses "entry" as the term for a parsing table. When reading in entry <name>, I would still not know which kind of a thing <name> is an entry in. entry is not very descriptive, but well, at least it is consistent

We inherit the terminology from [Camlp4](https://caml.inria.fr/pub/docs/tutorial-camlp4/tutorial003.html)/[Camlp5](https://camlp5.github.io/doc/html/grammars.html#a:Getting-started). It seems to be used with a similar meaning in the literature about parsing compiling (e.g. [here](https://books.google.de/books?id=05xA_d5dSwAC&pg=PA129&lpg=PA129&dq=grammar+entry+non-terminal&source=bl&ots=3PqxdCecKc&sig=MXlTorbZo4ikftReehaXSZABYPI&hl=fr&sa=X&ved=0ahUKEwjE_rP_uLHWAhUFuBoKHRezASsQ6AEIQzAE#v=onepage&q=grammar%20entry%20non-terminal&f=false)).

> We currently don't use that plugin because we have not yet figured out a way for users of Iris on Windows to get it.

This seems to be of general enough interest to justify an integration to the [existing](https://github.com/coq/coq/blob/master/plugins/syntax/string_syntax.ml) string syntax plugin. What do others think?

> (From what I hear, using opam is not an option there.)

Isn't it? To the risk of starting an emotional discussion, I would say that this would otherwise be a weak point of our opam-based communication...

Zimmi48 commented:
@herbelin Do you wish to get more feedback of this? If no, maybe we can remove the `needs: feedback` flag. If yes, do you want to have this PR discussed at the WG?

ejgallego commented:
@Zimmi48 I think the main point remaining of this PR and #873 is to be sure that it interacts well with the other wanted use case [tactic languages scope] I do indeed believe that the WG would be and ideal place to discuss and merge this.

herbelin commented:
>  Do you wish to get more feedback on this? If no, maybe we can remove the needs: feedback flag. If yes, do you want to have this PR discussed at the WG?

I don't need more feedback for this PR to be usable but I'd of course be happy to get more feedback so as to continue to fine-tune the design, especially regarding the `pat` feature (with or without a quote, and in particular `{pat|P}`, or form `{'pat|P}`?), and regarding the custom `entry` feature.

The typical questions are:
1. Do we provide the extended syntax `{pat|P}`, or rather `{'pat|P}`. The latter has the advantage to be consistent with the `'pat` syntax that we already find in binders. The former is lighter.
2. Do we adopt the syntax `in grammar entry foo`, or `in custom entry foo`, or `in entry foo`? (That would be more representative to get strictly more than one opinion.)
3. Which interaction between scopes and new entries? I don't have a clear idea myself yet, so maybe this is to be experimented first.

Answering question 1 is needed for a strict compatibility (currently, the defaut syntax for `sig` is `{pat|P}` which causes a few incompatibilities - 2 or 3 times in Travis). Depending on the question, I shall maybe implement some `id is constr as ident` or `id is constr as pattern` entry to better control the exact grammar rule one wants for binders.

> I think the main point remaining of this PR and #873 is to be sure that it interacts well with the other wanted use case [tactic languages scope] I do indeed believe that the WG would be and ideal place to discuss and merge this.

Sure.

herbelin commented:
@ppedrot:

New grammar entries are bound to new kinds of generic arguments, using an `ExtraArg`.

For intern/interp/subst, these new `ExtraArg` behave like `constr` ant it is only for parsing/printing, that they behave differently from `constr` and that it is justified to have a specific `ExtraArg`.

I could redirect the new entries to the `ExtraArg` of `constr` just after parsing so that they use the intern/interp/subst functions of `constr`. For that, it seems I would need a function of the form `is_constr_subentry_arg : ('a,'b,'c) Genarg.genarg_type -> (unit, (constr_expr, 'a) eq) union` to enforce the type constraint that the `wit` for the new entry has the same raw type `constr_expr` as `wit_constr`.

Or, I tell to declare the new entry not only for parsing but also for intern/interp/subst (`Genintern.register_intern0` etc.). These functions would be the same as for `constr`, but the problem then is that `intern_constr`/`interp_constr`/`subst_constr` are defined in the ltac plugin, and thus I would need that the grammar system declares a hook to be later instantiated by the ltac plugin.

Do you have an opinion on what solution to implement? Maybe the second approach is more consistent with the rationale of `wit`'s to be attached to a tuple of methods for parsing/interp/intern/subst/printing. Actually, I even don't know if I'm able to easily write such a type coercion function `is_constr_subentry_arg`.

RalfJung commented:
After talking with @ppedrot today, I finally understood that these are really not separate grammars -- but just new nonterminals in the same grammar. I did not realize that the entire grammar of Coq (well, what I consider to be "the main grammar" -- constr) is just one giant nonterminal with a gazillion rules. So I now understand why you did not like my names "table" or "grammar". "entry" is probably a better choice than either of the two.

However, I don't think "entry" is very understandable either. I don't have a good suggestion though.

herbelin commented:
Hi, I should be almost done with this PR. It is big but I would appreciate further comments, implementation-wise and feature-wise. I've found many bugs already and probably some remain.

Last steps I made are:
- support for using a custom entry in `Tactic Notation` as requested by Iris
- support for `constr as ident` to parse an `ident` as a `constr`, while the new default is that a `constr` is parsed as a `pattern` (this answer
- support for `strict pattern` to print a binding as a pattern only when it is not trivally a variable.
These two steps allow, as discussed at the last WG, to provide the notation `{'(x,y)|x+y=0}` and its variants in a way which preserve compatibility (i.e. `{'I:True|I=I}` and `{I:True|I=I}` now denote different terms, the same way as `fun I => I = I` and `fun 'I => I = I` denote different terms). 

This answers question 1 [above](https://github.com/coq/coq/pull/982#issuecomment-333193330). What still remains to do is a bit more of documentation. No change made on question 2. Still don't know how to address question 3 above.

robbertkrebbers commented:
@herbelin This is great. The next couple of weeks I'm pretty busy, but after that, I will try to experiment with the PR in Iris.

mattam82 commented:
I’m looking forward to playing with this too! I’m on the ‘pat side, « in
entry foo » looks fine to me, « in grammar of » maybe? I’m wondering if we
have to answer question 3 at all right now. Can’t one already use explicit
scope delimiters to make the two features interact?
I think you said yes at the WG, through an indirection token.
Le ven. 3 nov. 2017 à 17:00, Robbert Krebbers <notifications@github.com> a
écrit :

> @herbelin <https://github.com/herbelin> This is great. The next couple of
> weeks I'm pretty busy, but after that, I will try to experiment with the PR
> in Iris.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/coq/coq/pull/982#issuecomment-341746859>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAGARb4tlSO16boqbfbluPRYM7fqxOm4ks5syziYgaJpZM4O7day>
> .
>

herbelin commented:
An update/summary on providing `{pat|P}` or `{'pat|P}`:
- as already said, `{pat|P}` creates conflicts with existing `{C:T|P}` when `C` is a constructor, and indeed there is a difference of semantics between say `{I:nat|I=I}` and `{I:True|True}`;
- the notation `{ ' pat | P }` breaks the parsing of expressions of the form `{ forall x, P } + { Q }`; indeed the latter works thanks to a tolerance of Camlp5 in parsing `forall x, P` at level 200 while the rule asks to actually parse the interior of `{ ... }` at level 99 (the reason for 99 is to be below the rule for `M : T` which is at level 100, so that `{ x : A | P }` does not see `x : A` as a cast); adding an extra `"'"; pat = pattern` in parallel to `c = constr LEVEL "99"` apparently breaks the tolerance for `constr` at level 200;
- the notation `{' pat | P }` breaks notations for constr which starts with `'` as e.g. `'GL` in math-comp.

I'm out of ideas to support the notation in a compatible way, so I'm switching to a model where the notation is provided only on demand by requiring a special module `SpecifPatternNotations`.

herbelin commented:
Hi,

I moved the "custom entry" part to another PR (#6247). What remains in this PR is what is in the original message, i.e.:
- more general forms of recursive patterns are supported in recursive notations
- new modifers `x pattern`, `x strict pattern`, `x as pattern`, `x as strict pattern`, `x as ident`, `x at level n as pattern`, `x at level n as strict pattern`, `x at level n as ident` which can be used in notations to tell to parse a subterm as a pattern, or as a non-single-variable pattern, or, otherwise, for factorization purpose, as a constr, possibly at some given level, but nevertheless to be converted as a pattern, a strict pattern or a variable.

Then, it seems to me that this part of the PR starts to be pretty stable and I made the documentation.

Note that I finally found a hack to fix the `"{ ' pat | P }"` mentioned at item 2 [here](https://github.com/coq/coq/pull/982#issuecomment-346080849). So, it is available by default.

How to proceed?

I may try to further split this PR into smaller PRs as suggested by @ejgallego but I'm not fully sure that the inner commits all comply with Travis. So, there is a little risk that it becomes a pain.

There is also the question of granularity of the split. Some commits are small fix, some commits are incremental enhancements, some commits are pure redesign or cleaning of the code to help further enhancements, some are parts of other PRS (e.g. #978). The features properly speaking are few: support for `pattern` and more powerful recursive notations. Otherwise, there are typically the following small enhancements:

User messages enhancements:
- location and warnings when format disagree with the notation
- more precise explanation when a notation is not reversible for printing

Enhancements:
- support for "if t then pat then u else v" (a module to import)
- a better formatted printing for the unicode notations for binders
- support for `Locate "x + y"` in addition to `Locate "_ + _"`

PS: The [diff](https://github.com/herbelin/github-coq/compare/master+printing-factorized-match-clauses...master+notation-misc-extensions) after removal of #978 (still long though). I can answer questions.

JasonGross commented:
I'd like to see this in 8.8; is this likely?

ejgallego commented:
I am at least one person to blame as to slowing down @herbelin but sure I think as far as I know this is 100% in the 8.8 roadmap.

maximedenes commented:
> Let me strongly suggest that in the future please try to submit smaller and more self-contained PRs.

I cannot say enough how much I agree with this. I would prefer to merge 48 small PRs if they are independent, than one large. It's much easier to make sure nothing is blocked for no reason that way.

herbelin commented:
@maximedenes: How would you like it to be split? As said in my former comment, the features are few (support for entry pattern + miscellaneous small extensions of recursive notations). There are also a couple of not directly related other minor improvements. Would you prefer, say about 10 PRs, or rather say, around 5 PRs?

@ejgallego
> and also I am not very fond of the new printing options

What do you mean? Do you have something else in mind?

> There are some very interesting bits in this PR as the new registering API

Are you talking about the user-level API for the different variants of `pattern` (strict or not strict, parsed as a constr or as a pattern, etc.)? If yes, a look at the documentation of the new features might be the best way to have an overview of them (Section 12.1).

ejgallego commented:
> What do you mean? Do you have something else in mind?

I just think that we have too many options, and this is a problem sometimes.

Regarding the PR, again the problem is that this PR includes at least 10 logical changes. So bundling them together as 2 negative effects IMO: 1. a request for revision of one change blocks the merging of the other 9 changes. This is bad for the submitter. 2. The amount of bundling means that the PR becomes harder to understand and interesting bits / new possibilities go unnoticed by the other developers. This is indeed bad for other people interested in the work.

I think that is all, and there is no rule to follow here, but more of intuition. Like the merge/vs/rebase discussion, I think both make sense in different contexts.

herbelin commented:
> I just think that we have too many options, and this is a problem sometimes.

I'm actually not pushing for the options (but just to be sure, you are referring to the options from #978, on factorizing clauses). I have no idea whether users may want to use them or not. I made option preventively so that users feel comfortable with this change and that they can recover the way of doing they are familiar to.

> Regarding the PR, again the problem is that this PR includes at least 10 logical changes. So bundling them together as 2 negative effects IMO: 1. a request for revision of one change blocks the merging of the other 9 changes. This is bad for the submitter. 2. The amount of bundling means that the PR becomes harder to understand and interesting bits / new possibilities go unnoticed by the other developers. This is indeed bad for other people interested in the work.

I can try to split the PR into say 10 smaller PRs. But we need a protocol for updates, rebasings, etc. Since the PRs would depend the ones of the others, if I modify some individual PR, I would need to find a way to update the ones which depend on the modified PR. I was not very succesful yet in rebasing different PRs built one on top of another (as #978, this one, and #6247) and keep them synchronous. How do you do yourself to not redo the same rebase for all PRs that include the same commits? Is it with rerere?

An alternative could be that prepare links to a split of this PR in say 10 smaller diffs (as I did [here](https://github.com/herbelin/github-coq/compare/master+printing-factorized-match-clauses...master+notation-misc-extensions) above to remove the #978 part). For reviewing, this would behave the same I think as having different PRs. What do you think?

SkySkimmer commented:
>How do you do yourself to not redo the same rebase for all PRs that include the same commits? Is it with rerere?

I used to mess around with rebase --onto but the last one I did I just rebased the branch with the most commits then reset the other branch to the appropriate resulting commit.

ejgallego commented:
@herbelin sorry if I was not clear, IMHO this PR should be merged without a split. Not a point in slowing you down even more.

I was just trying to make a comment for _future_ PRs.

ejgallego commented:
This is of course my personal view based on how I tried to review / learn from your work.

I find it easier for example to discuss about the options in the PR that contains the change. Please **don't** take my opinion as any kind of rule or advice, it is just a very personal opinion.

JasonGross commented:
For rebasing, I generally run `git rebase -i` and delete any unwanted commits.  I suppose if there are 10 branches, you could write a script that, for each branch, in the correct order, does `git checkout $branch && (git rebase $base_branch || (git rebase --abort && git rebase -i $base_branch; exit 1)) || exit $?`

herbelin commented:
Well, thanks for `rebase` experience. I generally do myself a `git pull -r origin master` or a `git rebase -i origin/master`. I could indeed do a script sometimes, when it is clear that I would need it several times. About `-onto`, is it to ensure that, if I have PR B built on PR A, that the rebase done on A is reused litterally for the part of A which is in B? But then, do you combine it with `-i` to remove in advance the A which is in B (to avoid potential conflicts in applying the old A which is in PR B on top of the new PR A)?

herbelin commented:
@maximedenes: do you still prefer a split, and if yes of which kind, or do you subscribe to @ejgallego's proposal?

maximedenes commented:
Yeah sorry I wasn't clear. I also meant to insist on that for future PRs. So I do subscribe to @ejgallego's proposal.

SkySkimmer commented:
My old rebase process was (with `oldfoo` being some hash):
```
oldmaster -------------------------------> master
          \
           \
            \
             --> A ------> B

$ git rebase master A

oldmaster -------------------------------> master
          \                                      \
           \                                      \
            \                                      --> A
             --> oldA ------> B

$ git rebase --onto A oldA B

oldmaster -------------------------------> master
          \                                      \
           \                                      \
            \                                      --> A ------> B
             --> oldA ------> oldB (<- can be garbage collected)
```

My new process is (with `oldfoo` and `newfoo` hashes)
```
oldmaster -------------------------------> master
          \
           \
            \
             --> A ------> B

$ git rebase master B

oldmaster -------------------------------> master
          \                                      \
           \                                      \
            \                                      --> newA --------> B
             --> A ------> oldB

$ git checkout A
$ git reset --hard newA

oldmaster -------------------------------> master
          \                                      \
           \                                      \
            \                                      --> A ------> B
             --> oldA ------> oldB (<- can be garbage collected)
```

herbelin commented:
> git rebase --onto A oldA B

Ah, ok, no `-i` needed with that indeed.

And why did you change from rebase A then rebase B onto A, to rebase B then reset on A? Isn't it in fine the same elementary merge operations in the same order?

herbelin commented:
> Yeah sorry I wasn't clear. I also meant to insist on that for future PRs. So I do subscribe to @ejgallego's proposal.

Ok, noted. (And this may mean that I start to understand you better!)

SkySkimmer commented:
That way there's only 1 operation which requires attention (since reset always succeeds). Also it was somehow easier to find newA than oldA. Finally I don't know how to `--onto` with magit, but the other way I can do everything in magit.

herbelin commented:
@SkySkimmer: Thanks.

Zimmi48 commented:
Thanks for sharing your workflow @SkySkimmer. The `--onto` option is still needed when rebasing on someone else's rebased PR though ;-)

herbelin commented:
@ejgallego: I just saw that you wondered about benchmarking. Do you mean a special benchmarking for printing or a general benchmarking? (In any case, for a general benchmarking, I started #319.)

@maximedenes: I don't have a lot of time these days. Is it ok that I let you supervise the integration of this PR? Interacting further with @ejgallego if needed for the benchmarking, collecting the possible remaining questions about the PR, etc. I'll do a (last) rebase at the last minute. OK?

ejgallego commented:
The bench was just to be on the safe side as @JasonGross reported some trouble, @JasonGross , are the bench gonna cover your use case?

As for the rest I think we agreed that the PR could be merged as-is, so I suggest indeed to do a rebase once @maximedenes is gonna have time for a merge.

JasonGross commented:
@ejgallego Ah, probably not, if it uses `opam` rather than the CI.  I can update the opam target to build a couple of files that aren't installed to test this (change `[make "-j%{jobs}%"]` to `[make "-j%{jobs}%" "coq" "selected-specific-display"]`); shall I do this?

To be clear, @ejgallego , are you talking about https://github.com/coq/coq/issues/6416 ?  There the issue is https://github.com/coq/coq/pull/873 , and this seems unrelated (though I suppose it's better to be on the safe side).

ejgallego commented:
Hi @JasonGross , indeed I am talking about those bugs, so I thought it could be a good idea to have the bench system do some printing so we have a bit more safeguard against accidental performance regressions.

Indeed, printing [and parsing] is a hot spot in Coq due to the complexity class of many operations in the path, and these are far from sufficiently tested as of today IMHO.

herbelin commented:
The bench is [here](https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/319/console), ranging from +0.11 to -2.01. I may otherwise have missed the trouble reported by @jasongross, or at least I don't remember which it was.

maximedenes commented:
> @maximedenes: I don't have a lot of time these days. Is it ok that I let you supervise the integration of this PR? Interacting further with @ejgallego if needed for the benchmarking, collecting the possible remaining questions about the PR, etc. I'll do a (last) rebase at the last minute. OK?

Sure, I'll do my best. It may take a while given the various changes, but I'll make sure it doesn't stall.

JasonGross commented:
@herbelin I did not report any trouble with this PR.  I reported trouble with #873 resulting in around a 100x slowdown in printing

ppedrot commented:
@JasonGross FTR, the source of slowness of #873 can be mitigated by using a slightly less naive algorithm. I had starting writing a fix for it but I don't remember where I left it. If I have a bit of spare time I might come up with a solution before the freeze.

herbelin commented:
> Sure, I'll do my best. It may take a while given the various changes, but I'll make sure it doesn't stall.

Thanks!

herbelin commented:
@maximedenes: finally, I rebased today. The rebasing was actually only < 10 minutes, so not a big deal. However, two breakages took me 2-3 hours to be identified. One was due to an interaction with `DAst`: an exception was now raised lazily, thus escaping its handler and a `DAst.get` fixed it.

The second breakage is actually about a design choice to do, which was revealed by a test I introduced in the meantime. What to do with a `let` inbetween `fun`s or `forall`s (or more generally inbetween any sequence of binders). Should it be printed like `fun (a : A) (x := t) (b : B) => u`, or like `fun a : A => let x := t in fun b : B => u`?

In one of the commits, I experimented including the `let`s as part of sequencse of binders and, in general, this make the printing shorter. The problem is in case of overlapping with a notation starting with `let`, as in:
```coq
Notation foo y := (let x := 1 in x = y).
Check forall (y:nat), let x := 1 in x = y.
```
where one certainly wants `forall y : nat, foo y` rather than `forall (y : nat) (x := 1), x = y`. (See `Notations2.v` for a more realistic example.)

Of course, the overlapping of notations is not specific to `let`. It can happen in other situations as well. Ideally, one could try to detect overlapping but I'm a bit afraid to go this way, and, as a quick compromise, I stayed half-way, deciding to integrate inner `let`s to binding sequences but not trailing `let`s.

So, if anyone has some comment to do, don't hesitate. Otherwise, I believe I shall stick with the current situation.

maximedenes commented:
@herbelin do you consider this ready? If yes, maybe a last `rebase --whitespace=fix` and we can merge.

herbelin commented:
@maximedenes: as my last comment, I had to make a little regression on printing `let`-in-binders in the last version, but 1) noone reacted 2) Improve it is not prioritary in the short term. So, I guess this is ok as it is.

Zimmi48 commented:
The linter still complains. Did you use `--whitespace=fix` as suggested?

herbelin commented:
I did it and I think I commited what I did. I redid it and it changes nothing. The reported line (1083 in `constrextern.ml`) is not modified by the PR. However, it is at the border of a sequence of lines which is modified. Maybe the linter and `--whitespace=fix` have a different semantics there?

SkySkimmer commented:
When looking at the diff for the entire PR git thinks the line was modified, but when working per-commit like rebase it doesn't. I'll look into having the linter work per-commit.

Zimmi48 commented:
OK so in the meantime, this whitespace "error" should simply be ignored.

maximedenes commented:
I'm afraid this conflicts with #6745. Note sure which one I should give priority to. Both PRs are big. This one has been delayed a lot already, but the other one seems very tedious to rebase. So it probably depends how good @ejgallego 's tooling is. Any suggestion about the merge order, @ejgallego @herbelin ?

ejgallego commented:
IMHO this one has priority.

SkySkimmer commented:
Should #5585 be closed as the title says?

herbelin commented:
@skyskimmer: right, done.
